File: java/poker/src/main/java/Hand.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public  class Hand implements Comparable<Hand>{

    private final List<Card> cards;
    private final HandCategory handCategory;

    public Hand(String handString) {
        this.cards = Arrays.stream(handString.split("\\s+")).map(Card::new).toList();
        this.handCategory = Hand.categorizeHand(this.cards);
    }

    public List<Card> getCards() {
        return this.cards.stream().sorted(Comparator.reverseOrder()).toList();
    }

    HandCategory getHandCategory() {
        return this.handCategory;
    }

    static HandCategory categorizeHand(List<Card> cards) {

        List<Integer> sortedRanks = cards.stream().map(Card::getHankValue).sorted(Comparator.reverseOrder()).toList();

        boolean isStraight = sortedRanks.getFirst() - sortedRanks.getLast() == 4
                || (sortedRanks.getFirst() == 14 && (sortedRanks.get(1) - sortedRanks.getLast() == 3));

        boolean isFlush = cards.stream().map(Card::getSuit).distinct().count() == 1;

        String freq = cards.stream().map(Card::getHankValue)
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
                .values().stream().sorted(Comparator.reverseOrder())
                .map(Object::toString).collect(Collectors.joining());

        if (isStraight && isFlush) {
            return HandCategory.STRAIGHT_FLUSH;
        }

        if (freq.equals("41")) {
            return HandCategory.FOUR_OF_A_KIND;
        }

        if (freq.equals("32")) {
            return HandCategory.FULL_HOUSE;
        }

        if (isFlush) {
            return HandCategory.FLUSH;
        }

        if (isStraight) {
            return HandCategory.STRAIGHT;
        }

        return switch (freq) {
            case "311" -> HandCategory.THREE_OF_A_KIND;
            case "221" -> HandCategory.TWO_PAIR;
            case "2111" -> HandCategory.ONE_PAIR;
            default -> HandCategory.HIGH_CARD;
        };

    }

    @Override
    public int compareTo(Hand o) {
        return this.handCategory.getValue() - o.handCategory.getValue();
    }

    @Override
    public String toString() {
        return this.cards.stream().map(Card::toString).collect(Collectors.joining(" "));
    }
}

enum HandCategory {

    HIGH_CARD(0), ONE_PAIR(1), TWO_PAIR(2), THREE_OF_A_KIND(3), STRAIGHT(4),
    FLUSH(5), FULL_HOUSE(6), FOUR_OF_A_KIND(7), STRAIGHT_FLUSH(8);

    private final int value;

    HandCategory(int value) {
        this.value = value;
    }

    public int getValue() {
        return this.value;
    }
}

File: elixir/poker/lib/hand.ex
defmodule Hand do
  defstruct [:cards, :category]

  @categories %{
    straight_flush: 9,
    four_of_a_kind: 8,
    full_house: 7,
    flush: 6,
    straight: 5,
    three_of_a_kind: 4,
    two_pair: 3,
    one_pair: 2,
    high_card: 1
  }

  def parse_hand(hand) do
    cards = hand |> Enum.map(&Card.parse_card/1)

    %Hand{
      cards: cards,
      category: categorize_hand(cards)
    }
  end

  def sort_hands(hands) do
    hands
    |> Enum.sort(fn hand_0, hand_1 ->
      @categories[hand_0.category] >= @categories[hand_1.category]
    end)
  end

  def print_hand(hand) do
    hand.cards
    |> Enum.map(&Card.print_card/1)
  end

  defp categorize_hand(hand) do
    cond do
      straight?(hand) and flush?(hand) -> :straight_flush
      four_of_a_kind?(hand) -> :four_of_a_kind
      three_of_a_kind?(hand) and one_pair?(hand) -> :full_house
      flush?(hand) -> :flush
      straight?(hand) -> :straight
      three_of_a_kind?(hand) -> :three_of_a_kind
      two_pair?(hand) -> :two_pair
      one_pair?(hand) -> :one_pair
      true -> :high_card
    end
  end

  defp four_of_a_kind?(hand) do
    hand
    |> Enum.map(fn card -> card.rank end)
    |> Enum.frequencies()
    |> Map.values()
    |> Enum.member?(4)
  end

  defp three_of_a_kind?(hand) do
    hand
    |> Enum.map(fn card -> card.rank end)
    |> Enum.frequencies()
    |> Map.values()
    |> Enum.member?(3)
  end

  defp two_pair?(hand) do
    hand
    |> Enum.map(fn card -> card.rank end)
    |> Enum.frequencies()
    |> Map.values()
    |> Enum.sort(:desc)
    |> then(&(&1 == [2, 2, 1]))
  end

  defp one_pair?(hand) do
    hand
    |> Enum.map(fn card -> card.rank end)
    |> Enum.frequencies()
    |> Map.values()
    |> then(&(2 in &1))
  end

  defp straight?(hand) do
    sorted_hand = Card.sort_cards(hand) |> Enum.map(fn card -> card.value end)

    (Enum.at(sorted_hand, -1) == 2 and Enum.at(sorted_hand, 1) - Enum.at(sorted_hand, -1) == 3) or
      Enum.at(sorted_hand, 0) - Enum.at(sorted_hand, -1) == 4
  end

  defp flush?(hand) do
    hand
    |> Enum.map(fn card -> card.suit end)
    |> Enum.uniq()
    |> length() == 1
  end
end
